src/lib.rs

```diff
@@ -54,53 +54,64 @@ pub struct Cow<'a, T: Beef + ?Sized + 'a> {
 /// + `T::Owned` with `capacity` of `0` does not allocate memory.
 /// + `T::Owned` can be reconstructed from `*mut T` borrowed out of it, plus capacity.
 pub unsafe trait Beef: ToOwned {
-    /// Get capacity of owned variant of `T`. Return `None` for `0`.
-    /// Returning invalid capacity will lead to undefined behavior.
-    fn capacity(owned: &Self::Owned) -> Option<NonZeroUsize>;
-
-    /// Convert `&mut T::Owned` to `*mut T`, stripping `capacity`.
-    unsafe fn owned_ptr(owned: &mut Self::Owned) -> NonNull<Self>;
+    /// Convert `T::Owned` to `NonNull<T>` and capacity.
+    /// Return `None` for `0` capacity.
+    fn owned_into_parts(owned: Self::Owned) -> (NonNull<Self>, Option<NonZeroUsize>);
 
     /// Rebuild `T::Owned` from `NonNull<T>` and `capacity`. This can be done by the likes
     /// of [`Vec::from_raw_parts`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.from_raw_parts).
-    unsafe fn rebuild(ptr: NonNull<Self>, capacity: usize) -> Self::Owned;
+    unsafe fn owned_from_parts(ptr: NonNull<Self>, capacity: NonZeroUsize) -> Self::Owned;
 }
 
 unsafe impl Beef for str {
     #[inline]
-    fn capacity(owned: &String) -> Option<NonZeroUsize> {
-        NonZeroUsize::new(owned.capacity())
-    }
+    fn owned_into_parts(owned: String) -> (NonNull<str>, Option<NonZeroUsize>) {
+        // Convert to `String::into_raw_parts` once stabilized
+        let mut owned = mem::ManuallyDrop::new(owned);
+        let ptr = slice_from_raw_parts_mut(
+            owned.as_mut_ptr(),
+            owned.len(),
+        );
 
-    #[inline]
-    unsafe fn owned_ptr(owned: &mut String) -> NonNull<str> {
-        NonNull::new_unchecked(owned.as_mut_str() as *mut str)
+        (
+            unsafe { NonNull::new_unchecked(ptr as *mut str) },
+            NonZeroUsize::new(owned.capacity()),
+        )
     }
 
     #[inline]
-    unsafe fn rebuild(mut ptr: NonNull<Self>, capacity: usize) -> String {
+    unsafe fn owned_from_parts(mut ptr: NonNull<Self>, capacity: NonZeroUsize) -> String {
         String::from_utf8_unchecked(Vec::from_raw_parts(
             ptr.as_mut().as_mut_ptr(),
             ptr.as_mut().len(),
```

Comment:
If I'm being overcautious, I'd bring the `len` call up before the `as_mut_ptr`, and do it as `ptr.as_ref().len()`. If you want to be super correct, you can write `fn slice_len<T>(ptr: NonNull<[T]>) -> usize` to just directly extract the length metadata today: https://discordapp.com/channels/273534239310479360/592856094527848449/688117580040241162